ksqldb-engine/src/main/java/io/confluent/ksql/format/DefaultFormatInjector.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.format;
+
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.exception.ParseFailedException;
+import io.confluent.ksql.parser.properties.with.CreateSourceProperties;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.CreateSource;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.properties.with.CommonCreateConfigs;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.statement.Injector;
+import io.confluent.ksql.util.ErrorMessageUtil;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+import java.util.Optional;
+
+/**
+ * An injector which injects the key and value formats into the supplied {@code statement}.
+ *
+ * <p>The key format is only injected if:
+ * <ul>
+ * <li>The statement is a CT/CS.</li>
+ * <li>The statement does not specify the FORMAT property in its WITH clause.</li>
+ * <li>The statement does not specify the KEY_FORMAT property in its WITH clause.</li>
+ * </ul>
+ *
+ * <p>Similarly, the value format is only injected if the above conditions are met,
+ * where the KEY_FORMAT property is replaced with the VALUE_FORMAT property accordingly.
+ *
+ * <p>If any of the above are not true then the {@code statement} is returned unchanged.
+ *
+ * <p>This injector also validates that the KEY_FORMAT property is only supplied if the
+ * relevant feature flag is enabled.
+ */
+public class DefaultFormatInjector implements Injector {
+
+  public DefaultFormatInjector() {
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement
+  ) {
+    if (statement.getStatement() instanceof CreateSource) {
+      final ConfiguredStatement<CreateSource> createStatement =
+          (ConfiguredStatement<CreateSource>) statement;
+
+      if (createStatement.getStatement().getProperties().getKeyFormatName().isPresent()) {
+        throwIfKeyFormatDisabled(createStatement);
+      }
```

Comment:
> As discussed offline, this injector can't be a direct pass-through if the feature flag is not enabled because we still need to validate that VALUE_FORMAT is explicitly provided in this case

Agreed.

>  and I think it also makes sense to throw if KEY_FORMAT is supplied here rather than forcing downstream code to handle that.

I don't think we gain anything from checking here, if we're already checking in `EngineExecutor.plan`.  It feels like an unnecessary check. But it's no biggie, and having not played with the code, I may just be wrong ;)

Reply:
Ah, just remembered (rediscovered) the problem I ran into when I originally wanted the injector to be a pass-through if the feature flag isn't enabled: if the downstream (engine) code is updated to expect the statement contains a key format, and the injector doesn't add a key format because the feature flag is not enabled, then we fail downstream. 

Other options include:
- have the downstream code check the feature flag and default to KAFKA if the key format isn't present. This is undesirable because we'll be checking the feature flag in a bunch of different places. Even if we create a util method for it, we'll be adding a bunch of code for passing KsqlConfig into the util method which will be removed once the feature flag is removed, and doesn't seem worth the effort as a result.
- have CreateSourceProperties consume the feature flag and default to KAFKA if not set. This faces the same problem of CreateSourceProperties needing access to KsqlConfig which feels wrong.

The downside of having the injector always inject key format if not present, regardless of feature flag, is that when the feature flag is not set, users will still start seeing KEY_FORMAT appear in their statements which may be confusing. IMO this still seems less bad than the other options. WDYT?

> I don't think we gain anything from checking here, if we're already checking in EngineExecutor.plan. It feels like an unnecessary check.

The engine checks whether the key format is supported. This check is for whether the key format is supplied at all, and throws even if a valid (supported) key format is supplied, since users shouldn't be trying to supply key formats if the feature flag is not enabled. If we'd rather we can remove this check and allow users to supply `KEY_FORMAT='KAFKA'` even if the feature flag is not enabled. (My original reason for disallowing this was in an attempt to preserve the pre-feature flag behavior, but that's now been broken by the required change above.)

Reply:
> IMO this still seems less bad than the other options. WDYT?

Sounds good.

> If we'd rather we can remove this check and allow users to supply KEY_FORMAT='KAFKA' even if the feature flag is not enabled. 

I think it makes total sense. If we allow `KEY_FORMAT='KAFKA'` then its no longer a big weird that users will see `KEY_FORMAT` in their statements when ksqlDB displays them back to them.

Reply:
Cool. I've updated the PR to allow the use of `KEY_FORMAT` and `FORMAT` even when the feature flag is not enabled.